### PR TITLE
Improved the Client Trace Span generations in calling components APIs

### LINKED ISSUE #1484: Optimize the Client Trace Span generations in calling components APIs
Trace Server Span are generated at one place in the Dapr - each for gRPC/HTTP request.
However currently Client Span are generated for each http/gRPC api calls - such as GetState/PostState . We can improve and refactor the code in such a way that client spans are generated at one place before calling components.

One way is to create a middle layer/handler that every outbound components API calls are intercepted and invoked Client Span generation logic in Dapr side.

> /area runtime

### PR DESCRIPTION
**Pull request by shalabhms**
# Description

Trace Server Span are generated at one place in the Dapr - each for gRPC/HTTP request.
However currently Client Span are generated for each http/gRPC api calls - such as GetState/PostState . Improved code in such a way that client spans are generated at one place in middleware before calling components.

With changes , every outbound components API calls are intercepted and invoked Client Span generation logic in Dapr side at one place.

Removed the client span generations in each outbound API component calls.

## Issue reference

Closes #1484 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/diagnostics/http_tracing.go [hunk 1]
```diff
@@ -93,12 +109,28 @@ func addAnnotationsToSpan(req *fasthttp.Request, span *trace.Span) {
 	})
 }
 
+func isAppChannelServiceInvocationRequest(ctx *fasthttp.RequestCtx, appID, name string) bool {
```
Comment (yaron2) [inline pkg/diagnostics/http_tracing.go:112]: why do you need the fasthttp context here?

Comment (shalabhms) [reply pkg/diagnostics/http_tracing.go]: > why do you need the fasthttp context here?

Added earlier to take from userValues from reqCtx but not needed with change. I forgot to remove this param, have removed now. thanks.

File: pkg/diagnostics/http_tracing.go [hunk 2]
```diff
@@ -32,12 +32,28 @@ const (
...
+		if isAppChannelServiceInvocationRequest(appID, path) || isHealthzRequest(path) {
+			SpanContextToRequest(sc, &ctx.Request)
+			next(ctx)
+		} else {
+			spanName := fmt.Sprintf("%s:%s", method, path)
+
+			newCtx := NewContext((context.Context)(ctx), sc)
+			_, span := StartTracingClientSpanFromHTTPContext(newCtx, &ctx.Request, spanName, spec)
+			SpanContextToRequest(span.SpanContext(), &ctx.Request)
+			defer span.End()
```
Comment (youngbupark) [inline pkg/diagnostics/http_tracing.go:51]: I think we do not need defer here. instead we can move line 51 to between L55 and L56.

Comment (shalabhms) [reply pkg/diagnostics/http_tracing.go:51]: > I think we do not need defer here. instead we can move line 51 to between L55 and L56.

Agree , good catch !.

**Pull request merged**